src/streamlink/plugins/vimeo.py

```diff
@@ -8,70 +8,175 @@
+                    validate.optional("hls"): {"cdns": {validate.text: {"url": validate.any(None, validate.url())}}},
                     validate.optional("progressive"): validate.all(
-                        [{"url": validate.url(), "quality": validate.text}]
+                        [{"url": validate.url(), "quality": validate.text}],
                     ),
                 },
                 validate.optional("text_tracks"): validate.all(
-                    [{"url": validate.text, "lang": validate.text}]
+                    [{"url": validate.text, "lang": validate.text}],
                 ),
-            }
+            },
         },
     )
-    _player_schema = validate.Schema(
-        validate.transform(_config_re.search),
-        validate.any(None, validate.Schema(validate.get(1), _config_schema)),
-    )
 
+    def __init__(self, session, url):
+        self.session = session
+        self.url = url
+        self._auth_data = None
+        self._last_updated = None
+        self._updating = False
+
+    def get_player_data(self):
+        return self.session.http.get(self.url, schema=validate.Schema(
+            validate.transform(re.compile(r"var\s+config\s*=\s*({.+?})\s*;").search),
+            validate.any(None, validate.Schema(validate.get(1), VimeoAPI._config_schema)),
+        ))
+
+    def get_api_url(self):
+        return self.session.http.get(self.url, schema=validate.Schema(
+            validate.transform(re.compile(r'(?:"config_url"|\bdata-config-url)\s*[:=]\s*(".+?")').search),
+            validate.any(
+                None,
+                validate.Schema(
+                    validate.get(1),
+                    validate.parse_json(),
+                    validate.transform(html_unescape),
+                    validate.url(),
+                ),
+            ),
+        ))
+
+    def get_config_data(self, api_url):
+        return self.session.http.get(api_url, schema=VimeoAPI._config_schema)
+
+
+@pluginmatcher(re.compile(
+    r"https?://(player\.vimeo\.com/video/\d+|(www\.)?vimeo\.com/.+)"
+))
+class Vimeo(Plugin):
     arguments = PluginArguments(
         PluginArgument("mux-subtitles", is_global=True)
     )
 
-    def _get_streams(self):
-        if "player.vimeo.com" in self.url:
-            data = self.session.http.get(self.url, schema=self._player_schema)
+    def _get_data(self):
+        if "player.vimeo.com" in self.session.api.url:
+            return self.session.api.get_player_data()
         else:
-            api_url = self.session.http.get(self.url, schema=self._config_url_schema)
+            api_url = self.session.api.get_api_url()
             if not api_url:
                 return
-            data = self.session.http.get(api_url, schema=self._config_schema)
+            return self.session.api.get_config_data(api_url)
+
+    def _get_streams(self):
+        self.session.api = VimeoAPI(self.session, self.url)
+        data = self._get_data()
```

Comment:
There is no point in having the `_get_data()` method on the plugin class if it only calls other methods on the API instance. You are also referencing this method in `VimeoHLSStream._update_auth` above, which is bad. Move it to the VimeoAPI class.